docs/rfc/rfc-016-node-architecture.md

```diff
@@ -0,0 +1,83 @@
+# RFC 016: Node Architecture
+
+## Changelog
+
+- April 8, 2022: Initial draft (@cmwaters)
+- April 15, 2022: Incorporation of feedback
+
+## Abstract
+
+The `node` package is the entry point into the Tendermint codebase, used both by the command line and programatically to create the nodes that make up a network. The package has suffered the most from the evolution of the codebase, becoming bloated as developers clipped on their bits of code here and there to get whatever feature they wanted working.
+
+The decisions made at the node level have the biggest impact to simplifying the protocols within them, unlocking better internal designs and making Tendermint more intuitive to use and easier to understand from the outside. Work, in minor increments, has already begun on this section of the codebase. This document exists to spark forth the necessary discourse in a few related areas that will help the team to converge on the long term makeup of the node.
+
+## Discussion
+
+The following is a list of points of discussion around the architecture of the node:
+
+### Dependency Tree
+
+The node object is currently stuffed with every component that possibly exists within Tendermint. In the constructor, all objects are built and interlaid with one another in some awkward dance. My guiding principle is that the node should only be made up of the components that it wants to have direct control of throughout its life. The node is a service which currently has the purpose of starting other services up in a particular order and stopping them all when commanded to do so. However, there are many services which are not direct dependents i.e. the mempool and evidence services should only be working when the consensus service is running. I propose to form more of a hierarchical structure of dependents which forces us to be clear about the relations that one component has to the other. More concretely, I propose the following dependency tree:
+
+![node dependency tree](./images/node-dependency-tree.svg)
+
+Many of the further discussion topics circle back to this representation of the node.
+
+As pointed out by Michael (@creachadair), It's also important to distinguish two dimensions which may require different characteristics of the architecture. There is the starting and stopping of services and their general lifecycle management. What is the correct oder of operations to starting a node for example. Then there is the question of the needs of the service during actual operation. Does it require access to events? Which data does it need read access from and so forth.
```

Comment:
As a future note, try to make all your changes for a line in a single comment. I can't commit them if they're across several comments :)

Reply:
Sorry. I usually do, but these were enough unrelated that I didn't want you to feel like you had to adopt them all. I should have remembered it would screw up the later ones though.

Reply:
That's true. The other thing I could learn to do is break my paragraph up like you do in your RFC's

Reply:
> That's true. The other thing I could learn to do is break my paragraph up like you do in your RFC's

I think opinion is divided on that. I just use emacs's `fill-paragraph`, but some folks don't like how that causes ragged ends at different screen widths.

I'll just be more careful about that detail next time 😁